### DISCUSSION TITLE
learning  $dnstype modifier

### DISCUSSION THREAD

**Discussion by gentlyxu** [AdguardTeam/AdGuardHome#3069] [Q&A] [answered]
### Issue Details

* **Version of AdGuard Home server:** 
  * v0.106.1
* **How did you install AdGuard Home:** 
  *  Github releases
* **How did you setup DNS configuration:** 
  * Router
* **If it's a router or IoT, please write device model:** 
  * Raspberry Pi 4
* **CPU architecture:** 
  * ARM
* **Operating system and version:** 
  * Linux raspberrypi 5.4.51-v7l+

### Expected Behavior

Do not block the domain "www.iconfont.cn"

### Actual Behavior

blocked

### Screenshots

<details><summary>Screenshot:</summary>

<img width="910" alt="blocked1" src="https://user-images.githubusercontent.com/1243610/116769654-38be8080-aa70-11eb-920a-02e5b8959105.png">

</details>

### Additional Information

The old rule I use are:
```
/^(\S+\.)?advert/
```
I want to block all domain names with 'advert' in keywords. Someone reported to me that the domain "www.iconfont.cn" was mishandled -- https://github.com/privacy-protection-tools/anti-AD/issues/410 

So I updated the rule as follow:

```
/^(\S+\.)?advert/$dnstype=~CNAME,important
```
I want adgh to exclude CNAME queries, But it is not working now, CNAME query is still blocked(see the screenshot)

So how can I solve the problem?

**Comment by banbendalao** [comment]
When I use `/^(\S+\.)?advert/$dnstype=CNAME`, `www.iconfont.cn` will not be blocked. This may be a BUG

**Comment by ainar-g** [comment]
Hello!  Wouldn't adding an exception to filtering solve the issue?  That is:

```none
@@www.iconfont.cn^
```

**Reply by gentlyxu** [reply]
> Hello! Wouldn't adding an exception to filtering solve the issue? That is:
> 
> ```
> @@www.iconfont.cn^
> ```

Of course it can... but what I want to know is why `$dnstype` does not take effect?

**Comment by ameshkov** [comment]
Wouldn't it be better to use `denyallow`?

`/^(\S+\.)?advert/$denyallow=www.iconfont.cn`

**Comment by ameshkov** [comment, marked as answer]
Ah, I see, it blocks a domain because it points to `na61-na62.wagbridge.advertisement.inner.alibabacorp.com` and `na61-na62.wagbridge.advertisement.inner.alibabacorp.com.gds.alibabadns.com ` which makes sense -- your rule matches both domains.

There're two options to fix this.

1. Unblock those two domains. This is okay if they're only used as CNAME targets.
```
@@||wagbridge.advertisement.inner.alibabacorp.com^
@@||wagbridge.advertisement.inner.alibabacorp.com.gds.alibabadns.com^
```

2. Use `denyallow` for the same reason:
```
/^(\S+\.)?advert/$denyallow=alibabacorp.com|alibabadns.com
```

Regarding `dnstype` - the query in the original issue is of type `A` so `$dnstype=CNAME` simply does nothing to it.

**Reply by gentlyxu** [reply]
so, I misunderstood the usage of `$dnstype`. Thanks for your explanation

I will try option 2....very cool
